File: parameters_py/parameters_py/param_test.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import rclpy
from rclpy.node import Node

class MiniParam(Node):
    def __init__(self, name):
        super().__init__(name)
        # set param description
        from rcl_interfaces.msg import ParameterDescriptor
        parameter_descriptor = ParameterDescriptor(description='This parameter is mine!')
        # "set in python !!" will be cover by launch file: "set in launch !!"
        self.declare_parameter("demo_param", "set in python !!", parameter_descriptor)
        self._timer = self.create_timer(1, self.timer_cb)

    def timer_cb(self):
        my_param = self.get_parameter("demo_param").get_parameter_value().string_value
        self.get_logger().info("param: %s" % my_param)

        # Reset all parameters to prevent external modifications
        all_parameters = []
        demo_param = rclpy.parameter.Parameter(
            'demo_param',
            rclpy.Parameter.Type.STRING,
            'set in python !!'
        )
        all_parameters.append(demo_param)
        self.set_parameters(all_parameters)
            
            
def main(args=None):
    rclpy.init(args=args)
    node = MiniParam("test_param_py")
    rclpy.spin(node)
    
if __name__ == '__main__':
    main()

File: parameters_cpp/src/param_test.cpp
#include <chrono>
#include <functional>
#include <string>

#include "rclcpp/rclcpp.hpp"

class MiniParam : public rclcpp::Node {
public:
    MiniParam() : Node("test_param_cpp") {
        // set param description
        auto param_desc = rcl_interfaces::msg::ParameterDescriptor{};
        param_desc.description = "This parameter is mine!";
        // "set in cpp !!" will be cover by launch file: "set in launch !!"
        this->declare_parameter("demo_param", "set in cpp !!", param_desc);
        timer_ = this->create_wall_timer(std::chrono::seconds(1), std::bind(&MiniParam::timer_cb, this));
    }

    void timer_cb() {
        std::string demo_param = this->get_parameter("demo_param").as_string();
        RCLCPP_INFO(this->get_logger(), "demo_param: %s", demo_param.c_str());

        // Reset all parameters to prevent external modifications
        std::vector<rclcpp::Parameter> all_parameters;
        all_parameters.push_back(rclcpp::Parameter("demo_param", "set in cpp !!"));
        this->set_parameters(all_parameters);
    }

private:
    rclcpp::TimerBase::SharedPtr timer_;

};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);

    rclcpp::spin(std::make_shared<MiniParam>());

    rclcpp::shutdown();

    return 0;
}